Add WebSocket generator for real-time LLM security testing

dyrtyData opened the pull request:
## Summary

This PR adds a WebSocket generator to enable garak to test WebSocket-based LLM services for security vulnerabilities. Many modern chat-based LLM services use WebSocket connections for real-time bidirectional communication, and this generator extends garak's testing capabilities to cover these services.

**Key Features:**
- Full WebSocket protocol support using the professional `websockets` library
- Multiple authentication methods: Basic Auth, Bearer tokens, and custom headers
- Template-based messaging aligned with REST generator patterns (`req_template`, `req_template_json_object`)
- JSON response extraction with JSONPath support
- Typing indicator handling for chat-based LLMs
- Smart detection and graceful skipping of unsupported scenarios (system prompts, multi-turn conversations)
- Configurable timeouts, SSL verification, and connection management

## Files Changed

- `garak/generators/websocket.py` - Main WebSocket generator implementation (~390 lines)
- `docs/source/garak.generators.websocket.rst` - Comprehensive RST documentation with examples
- `docs/source/generators.rst` - Added websocket generator to documentation index
- `tests/generators/test_websocket.py` - Complete test suite with 17+ tests
- `pyproject.toml` - Added `websockets>=13.0` dependency
- `requirements.txt` - Added `websockets>=13.0` dependency

## Usage Example

```bash
# Basic usage with echo server
garak --model_type websocket.WebSocketGenerator \
      --generator_options '{"websocket": {"WebSocketGenerator": {"uri": "wss://echo.websocket.org"}}}' \
      --probes dan

# With authentication (password via environment variable)
export WEBSOCKET_API_KEY="your_secure_password"
garak --model_type websocket.WebSocketGenerator \
      --generator_options '{"websocket": {"WebSocketGenerator": {"uri": "ws://localhost:3000", "auth_type": "basic", "username": "user"}}}' \
      --probes encoding
```

## Configuration Options

| Parameter | Default | Description |
|-----------|---------|-------------|
| `uri` | `wss://echo.websocket.org` | WebSocket URL (ws:// or wss://) |
| `auth_type` | `none` | Authentication: none, basic, bearer, custom |
| `username` | None | Basic auth username |
| `api_key` | None | Bearer token or basic auth password |
| `req_template` | `$INPUT` | String template with $INPUT, $KEY, $CONVERSATION_ID |
| `req_template_json_object` | None | JSON object template |
| `response_json` | `false` | Parse responses as JSON |
| `response_json_field` | `text` | JSON field to extract (supports JSONPath: `$.data.message`) |
| `response_after_typing` | `true` | Wait for typing indicators to complete |
| `typing_indicator` | `typing` | String to detect typing status |
| `request_timeout` | 20 | Seconds to wait for response |
| `connection_timeout` | 10 | Seconds to wait for connection |
| `verify_ssl` | `true` | SSL certificate verification |

## Verification

- [x] All CI tests pass (Linux, macOS, Windows across Python 3.10, 3.12, 3.13)
- [x] WebSocket-specific tests: 17 passed, 3 skipped
- [x] Generator integration tests pass
- [x] CLI functionality verified working
- [x] DCO signed commits
- [x] Documentation follows garak RST patterns
- [x] Code follows garak generator patterns and conventions

### Manual Testing
- [x] Tested against echo.websocket.org public service
- [x] Tested against custom WebSocket LLM service with authentication
- [x] SSH tunnel compatible for secure remote testing

## Review History

This PR went through extensive review with maintainers @jmartin-tech and @leondz over 50+ commits:

**Key improvements from review feedback:**
1. Migrated from custom socket code to professional `websockets` library
2. Aligned template configuration with REST generator patterns
3. Implemented JSON response extraction with JSONPath support
4. Added `DEFAULT_CLASS` for module-level plugin discovery
5. Fixed security: removed dangerous URI scheme fallback, sanitized logging
6. Moved environment variable access to `_validate_env_var` method
7. Standardized on `api_key` for all authentication (no password in configs)
8. Added smart detection for unsupported scenarios (system prompts, multi-turn)
9. Migrated to use `_has_single_turn` pattern from base Generator

## Known Limitations

- **Single-turn only**: WebSocket generator gracefully skips multi-turn conversations (logs warning)
- **No system prompts**: System role messages are detected and skipped with warning
- **Session paradigm**: WebSocket session management differs from stateless HTTP; complex conversation flows may need future work as real-world WebSocket LLM services emerge

## Security Considerations

- No sensitive data logged (prompt content sanitized)
- Credentials stored via environment variables only (`WEBSOCKET_API_KEY`)
- SSL verification enabled by default
- No dangerous URI scheme fallbacks (HTTP -> WebSocket conversion removed)

## garak impacts

Resolves #435

dyrtyData commented on docs/websocket_generator.md:
Fixed in commit e220924b5e0738bb355214c3302a232b9c9dd333

dyrtyData commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,355 @@
+        api_key=api_key,
+        conversation_id=conversation_id,
+        **kwargs
+    )
```
Fixed in commit e220924b5e0738bb355214c3302a232b9c9dd333

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,410 @@
+        try:
+            # Send message
+            await self.websocket.send(message)
+            logger.debug(f"Sent message: {message[:100]}...")
```
This is logging raw content of intentionally malicious prompts that may contain characters that could cause issues or even security impacts for active log watchers.

leondz commented on garak/generators/websocket.py:
`DEFAULT_CLASS` needs to be defined for this module

dyrtyData replied:
Fixed in 0bb5386c. Added `DEFAULT_CLASS = "WebSocketGenerator"` to enable loading via `--model_type websocket`. The module now follows the same pattern as other generators (langchain, test, rasa, etc.).

leondz commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,379 @@
+
+    DEFAULT_PARAMS = {
+        "uri": "wss://echo.websocket.org",
+        "name": "WebSocket LLM",
```
this needs to come out and become a class-level immutable attribute. running the tests will indicate the current failure

leondz commented on docs/source/garak.generators.websocket.rst:
```diff
@@ -0,0 +1,225 @@
+Uses the following options from ``_config.plugins.generators["websocket"]["WebSocketGenerator"]``:
+
+* ``uri`` - the WebSocket URI (ws:// or wss://); can also be passed in --model_name
+* ``name`` - a short name for this service; defaults to "WebSocket LLM"
```
not configurable

leondz commented on docs/source/garak.generators.websocket.rst:
```diff
@@ -0,0 +1,225 @@
+* ``response_json`` - is the response in JSON format? (bool)
+* ``response_json_field`` - which field contains the response text? Supports JSONPath (prefix with ``$``)
+* ``response_after_typing`` - wait for typing indicators to complete? (bool)
+* ``typing_indicator`` - string that indicates typing status; default "typing"
```
can you say more about this? comment is insufficiently specific

dyrtyData commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,379 @@
+from typing import List, Union, Dict, Any, Optional
+from urllib.parse import urlparse
+import websockets
+from websockets.exceptions import ConnectionClosed, WebSocketException
```
Fixed in 35614feb. Removed unused `WebSocketException` import from `websockets.exceptions` (line 16). Only `ConnectionClosed` is actually used in the exception handling (line 293).

dyrtyData commented:
> Wasn't able to get this example running:
> 
> Instantiation of the plugin fails after the first request
> 
> ```
>  python -m garak --model_type websocket --model_name WebSocketGenerator   --generator_options '{"websocket": {"WebSocketGenerator": {"endpoint": "ws://echo.websocket.org"}}}'
> garak LLM vulnerability scanner v0.13.1.pre1 ( https://github.com/NVIDIA/garak ) at 2025-10-29T08:42:07.455495
> 📜 logging to /home/lderczynski/.local/share/garak/garak.log
> attribute name must be string, not 'type'
> ```
> 
> Recommend running the tests and running the examples provided, and getting all to pass

Fixed in 0bb5386c. Changed line 79 from `super().__init__(self.name, config_root)` to `super().__init__("WebSocket Generator", config_root)` to prevent the TypeError.

The name is now properly initialized as a class-level immutable attribute that can still be overridden via config or CLI (per the discussion on name configurability).

Instantiation now works correctly:
```bash
python -m garak --model_type websocket --probes test.Test \
  --generator_options '{"websocket": {"WebSocketGenerator": {"uri": "wss://echo.websocket.org"}}}'
  ```


All tests passing. Ready for review.

dyrtyData commented:
@jmartin-tech  or @leondz 
Could you please re-run the failed jobs? Both failures are network timeouts to HuggingFace, not code issues:

  Linux (ubuntu-24.04-arm, 3.10):
  FAILED tests/detectors/test_detectors.py::test_detector_detect[detectors.packagehallucination.JavaScriptNpm]
  requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='huggingface.co', port=443): Read timed out

  macOS (3.13):
  FAILED tests/detectors/test_detectors.py::test_detector_detect[detectors.packagehallucination.PythonPypi]
  requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='huggingface.co', port=443): Read timed out

  Thanks!

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,389 @@
+        except Exception as e:
+            logger.error(f"WebSocket generation failed: {e}")
+            message = Message(text="")
+            return [message] * min(generations_this_call, 1)
```
Same:
```suggestion
            return [message]
```

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,389 @@
+
+        except Exception as e:
+            logger.error(f"Failed to connect to WebSocket {self.uri}: {e}")
+            raise
```
This is simply logging to re-raising, would this gaining something by with a stricter scope?

leondz commented on garak/generators/websocket.py:
```diff
@@ -382,7 +416,7 @@ def __del__(self):
                 loop.close()
-            except:
+            except Exception:
                 pass  # Ignore cleanup errors
```
recommend catching exception msg and logging at warning level

```
except Exception as e:
        logging.warning("websocket teardown error:", exc_info=e)

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,423 @@
+            logger.warning(
+                "Failed to extract JSON field '%s': %s, returning raw response",
+                self.response_json_field,
+                e,
```
```suggestion
                exc_info=e
```

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,423 @@
+                if len(responses) == 1:
+                    return str(responses[0].value)
+                elif len(responses) > 1:
+                    return str(responses[0].value)
```
Why two conditions if always only returns a single entry?

```suggestion
                if len(responses) >= 1:
                    return str(responses[0].value)
```

or `join` with newlines:

```suggestion
                if len(responses) == 1:
                    return str(responses[0].value)
                elif len(responses) > 1:
                    return str("\n".join(response.value for response in responses))
```

jmartin-tech commented on garak/generators/websocket.py:
```diff
@@ -0,0 +1,423 @@
+                "Failed to extract JSON field '%s': %s, returning raw response",
+                self.response_json_field,
+                e,
+            )
```
```suggestion
            logger.warning(
                "Failed to extract JSON field '%s' returning raw response",
                self.response_json_field,
                exc_info=e,
            )
```
